File: onecard-server/src/modules/game/inference/action-mask.ts
import type { GameState } from '@/modules/game/domain/types/gameState';
import type { PokerCardPropsWithId } from '@/modules/game/domain/types/pokerCard';
import { isValidPlay } from '@/modules/game/domain/utils/cardUtils';

export const buildActionMask = (
  state: GameState,
  maxHandSize: number,
): boolean[] => {
  const mask: boolean[] = Array.from({ length: maxHandSize + 1 }, () => false);

  const hand: PokerCardPropsWithId[] = state.players[0]?.hand ?? [];

  if (!state.discardPile.length) {
    for (let i = 0; i < hand.length && i < maxHandSize; i++) {
      mask[i] = true;
    }
    mask[maxHandSize] = hand.length < maxHandSize;
    return mask;
  }

  const topCard = state.discardPile[0];
  const damage = state.damage;

  for (let i = 0; i < maxHandSize; i++) {
    if (i >= hand.length) {
      mask[i] = false;
      continue;
    }
    mask[i] = isValidPlay(hand[i], topCard, damage);
  }

  // 드로우는 손패가 maxHandSize에 도달하지 않았을 때만 허용
  mask[maxHandSize] = hand.length < maxHandSize;

  return mask;
};

export const applyActionMask = (
  logits: number[],
  mask: boolean[],
): number[] => {
  if (logits.length !== mask.length) {
    throw new Error(
      `logits length ${String(logits.length)} and mask length ${String(mask.length)} mismatch`,
    );
  }
  const NEG_INF = -1e9;
  return logits.map((v, idx) => (mask[idx] ? v : NEG_INF));
};

export const selectAction = (maskedLogits: number[]): number => {
  // argmax 선택; 필요하면 샘플링 구현 가능
  let bestIdx = 0;
  let bestVal = maskedLogits[0];
  for (let i = 1; i < maskedLogits.length; i++) {
    if (maskedLogits[i] > bestVal) {
      bestVal = maskedLogits[i];
      bestIdx = i;
    }
  }
  if (bestVal === -Infinity || bestVal <= -1e8) {
    throw new Error('No valid action after masking');
  }
  return bestIdx;
};

export interface EngineActionPayload {
  type: 'PLAY_CARD' | 'DRAW_CARD';
  playerIndex?: number;
  cardIndex?: number;
  amount?: number;
}

export const mapActionIndexToPayload = (
  actionIndex: number,
  state: GameState,
  maxHandSize: number,
): EngineActionPayload => {
  const hand = state.players[0]?.hand ?? [];
  if (actionIndex === maxHandSize) {
    return { type: 'DRAW_CARD', amount: 1 };
  }
  if (actionIndex < 0 || actionIndex >= maxHandSize) {
    throw new Error(`actionIndex ${String(actionIndex)} out of bounds`);
  }
  if (actionIndex >= hand.length) {
    // 마스킹이 제대로 되었다면 여기 오지 않아야 함
    throw new Error(
      `actionIndex ${String(actionIndex)} exceeds hand length ${String(hand.length)}`,
    );
  }
  return { type: 'PLAY_CARD', playerIndex: 0, cardIndex: actionIndex };
};

File: onecard-server/src/onecard_api/inference/action_mask.py
from __future__ import annotations

from typing import TypedDict

from onecard_api.domain.card_utils import is_valid_play
from onecard_api.domain.types import GameState, PokerCard


class EngineActionPayload(TypedDict, total=False):
    type: str
    playerIndex: int
    cardIndex: int
    amount: int


def build_action_mask(state: GameState, max_hand_size: int) -> list[bool]:
    mask = [False] * (max_hand_size + 1)
    hand: list[PokerCard] = state.get("players", [{}])[0].get("hand", []) if state.get("players") else []

    if not state.get("discardPile"):
        for i in range(min(len(hand), max_hand_size)):
            mask[i] = True
        mask[max_hand_size] = len(hand) < max_hand_size
        return mask

    top_card = state["discardPile"][0]
    damage = state.get("damage", 0)

    for i in range(max_hand_size):
        if i >= len(hand):
            mask[i] = False
            continue
        mask[i] = is_valid_play(hand[i], top_card, damage)

    mask[max_hand_size] = len(hand) < max_hand_size
    return mask


def apply_action_mask(logits: list[float], mask: list[bool]) -> list[float]:
    if len(logits) != len(mask):
        raise ValueError(
            f"logits length {len(logits)} and mask length {len(mask)} mismatch"
        )
    neg_inf = -1e9
    return [v if mask[idx] else neg_inf for idx, v in enumerate(logits)]


def select_action(masked_logits: list[float]) -> int:
    best_idx = 0
    best_val = masked_logits[0]
    for idx, val in enumerate(masked_logits[1:], start=1):
        if val > best_val:
            best_idx = idx
            best_val = val
    if best_val <= -1e8:
        raise ValueError("No valid action after masking")
    return best_idx


def map_action_index_to_payload(
    action_index: int, state: GameState, max_hand_size: int
) -> EngineActionPayload:
    hand: list[PokerCard] = state.get("players", [{}])[0].get("hand", []) if state.get("players") else []
    if action_index == max_hand_size:
        draw_amount = max(1, int(state.get("damage", 0)))
        return {"type": "DRAW_CARD", "amount": draw_amount}

    if action_index < 0 or action_index >= max_hand_size:
        raise ValueError(f"actionIndex {action_index} out of bounds")
    if action_index >= len(hand):
        raise ValueError(
            f"actionIndex {action_index} exceeds hand length {len(hand)}"
        )
    return {"type": "PLAY_CARD", "playerIndex": 0, "cardIndex": action_index}
